File: Api/src/main/java/com/horecastock/api/model/Producto.java
package com.horecastock.api.model;

import com.fasterxml.jackson.annotation.JsonIgnore;
import jakarta.persistence.*;
import jakarta.validation.constraints.*;
import lombok.*;

import java.util.List;
import java.util.UUID;

@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
@Entity
@Table(name = "productos")
public class Producto {

    @Id
    @GeneratedValue
    private UUID id;

    @NotBlank(message = "El nombre es obligatorio")
    @Size(min = 2, max = 100, message = "El nombre debe tener entre 2 y 100 caracteres")
    @Column(nullable = false, length = 100)
    private String nombre;

    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "categoria_id", nullable = false)
    private Categoria categoria;

    @Column(unique = true, length = 50)
    private String codigoBarras;

    @NotNull(message = "El stock actual es obligatorio")
    @Min(value = 0, message = "El stock actual no puede ser negativo")
    @Column(nullable = false)
    private Double stockActual;

    @NotNull(message = "El stock mínimo es obligatorio")
    @Min(value = 0, message = "El stock mínimo no puede ser negativo")
    @Column(nullable = false)
    private Double stockMinimo;

    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "unidad_medida_id", nullable = false)
    private UnidadMedida unidadMedida;

    @NotNull(message = "El precio por unidad es obligatorio")
    @DecimalMin(value = "0.0", inclusive = false, message = "El precio debe ser mayor que 0")
    @Column(nullable = false)
    private Double precioUnidad;

    @Builder.Default
    @Column(name = "activo")
    private Boolean activo = true;

    // Historial de movimientos de stock de este producto
    @OneToMany(mappedBy = "producto", cascade = CascadeType.ALL, orphanRemoval = true)
    @JsonIgnore
    private List<MovimientoStock> movimientosStock;
}

File: stockly/lib/models/producto.dart
// Modelo que representa un producto del inventario.
// El método fromJson convierte el JSON de la API a un objeto Dart.
class Producto {
  final String id;
  final String nombre;
  final String? codigoBarras;
  final double stockActual;
  final double stockMinimo;
  final double? precioUnidad;
  final bool activo;
  final String? categoriaNombre;
  final String? unidadMedidaNombre;

  Producto({
    required this.id,
    required this.nombre,
    this.codigoBarras,
    required this.stockActual,
    required this.stockMinimo,
    this.precioUnidad,
    required this.activo,
    this.categoriaNombre,
    this.unidadMedidaNombre,
  });

  factory Producto.fromJson(Map<String, dynamic> json) {
    return Producto(
      id: json['id'] as String,
      nombre: json['nombre'] as String,
      codigoBarras: json['codigoBarras'] as String?,
      stockActual: (json['stockActual'] as num).toDouble(),
      stockMinimo: (json['stockMinimo'] as num).toDouble(),
      precioUnidad: json['precioUnidad'] != null
          ? (json['precioUnidad'] as num).toDouble()
          : null,
      activo: json['activo'] as bool? ?? true,
      categoriaNombre: json['categoria']?['nombre'] as String?,
      unidadMedidaNombre: json['unidadMedida']?['nombre'] as String?,
    );
  }

  // Devuelve true si el stock está por debajo del mínimo
  bool get bajominimo => stockActual < stockMinimo;
}
